README.md
-- Audinotes
Date: Fri, 03/02/2023
Author: Coolbrother
Multitrack audio recorder.


changelog
-- changelog for audinotes
Date: Fri, 03/02/2023
Last update: Thu, 16/02/2023

Date: Thu, 16/02/2023
-- Version 0.3: Good Work
-- Adding: (bpm command to set bpm) and updating the Help list with new commands

-- Adding: (replace_data, merge_data functions in AudiPlayer class for factorizing
arrange_track function) 

-- Updating: changing audio buffersize from 1024 to 256
for better performance, and less latency, and adding capability for shifting
record track in (arrange_track function) 

-- Updating: (audiportdriver.py module)
-- Adding: (get_input_latency, get_output_latency function in audiportdriver
class, and inserting latency evaluation with shift_samples in arrange_track
function in AudiPlayer class) 

-- Updating: (audinotes module to inserting auditools
module for gen_notes function) 

-- Updating: (AudiNotes class with separate AudiTrack class) 
-- Adding: (toggle_mute function in AudiTrack, AudiPlayer classes, and commands 'x'
in MainApp class to toggle the muted track) 

-- Updating: (metronome synchro with
independant length and looping) 

-- Adding: (init_track from AudiPlayer class, and
test function from MainApp class) 
Updating: (_gen_notes function to pass
arguments, and arrange_track, _audio_callback functions to record in looping
mode) 

-- Adding: (debug function in audinotes.py module), and updating (for loop
to iterate on two channels in AudiTrack, AudiMetronome classes) 
Adding: (Mod 0,
for static array mode, Mode 2: for dynamic mode, init_click, gen_click,
get_position, set_position functions in the AudiMetronome class), and adding:
(Synchronisation between Current Track and the Metronome in _audio_callback
function with start_playing and start_clicking state) 

-- Adding: (is_looping,
set_looping, toggle_loop functions, in AudiTrack class, and command: 'l' in
MainApp class) 

-- Adding: (_mid2freq, _gen_notes, function), for testing of
synchro with the metronome 
Adding file: (audimetronome.py to manage
AudiMetronome class) 
-- Adding (Not Compile): (clear_screen function, portable for
Windows and Linux plateform, in MainApp class), and removed (AudioMetronome in
audionotes module) 

-- Adding: (get_bpm, set_bpm, gen_click, start_click, stop_click, in AudiMetronome class, and get_bpm,
set_bpm, toggle_click, functions in AudiPlayer class, and 'bpmi', 'bpmd', as
commands in MainApp class) Adding: (AudiMetronome class with static array for
tone beat, and dynamic silence length for blank beat) 

-- Updating: (arrange_track
function) using (the deque appendleft method instead of the List.append
method), for really better performance. 

-- Updating: (arrange_track function) to
resets player position after recording 

-- Adding: (set_arm_muted, is_arm_muted
functions) and updating (start_record, stop_record functions) to improve
recording mode. 

-- Adding: (AudiTrack class, and updating audio_callback and
arrange_track functions to work with it) 
#------------------------------------------------------------------------------

Last update: Sat, 04/02/2023
-- Version 0.2: Lite and Simple
-- Adding: (get_state, toggle_record, notify functions to the AudiPlayer
class), and adding commands (sta, status to the MainApp class))
-- Adding: (MainApp class for interface managing)

Last update: Fri, 03/02/2023
-- Version: 0.1
-- Adding:
		portdriver class for audio managing
		Player class for managing audio tracks.
		Commandline Interactive for functions.
		Adding functions:
		play, pause, stop, start_record, stop_record, toggle_rec_mode, recwire,
		arrange_track,get_position, set_position, get_length, rewind, forward,  goto_start, goto_end,
		is_running, is_playing, is_paused, is_recording,
		samples_to_sec.
		Adding help for commands.
		Adding readline for history and bash commands.


Last update: Wed, 01/02/2023
Version: 0.1
-- Test unlimited recording on PyAudio
Inspired from playrec module.
Note: Work fine.
with: output_device_index =15: default output Soundcard with only output
input_device_index =0: default Input Soundcard, with only input
output_device_index =6: Eternal output Soundcard
input_device_index =6: External input Soundcard
        
Date: samedi, 26/07/14 02/43/56
Author: Coolbrother




src/audimixer.py
#!/usr/bin/python3
"""
    File: audimixer.py
    Mixer V2
    Lite Mixer klass to manage audio drivers
    Date: Thu, 16/02/2023
    Author: Coolbrother
"""

import audiportdriver as aupor

def beep():
    print("\a")

#-------------------------------------------

class AudiMixer(object):
    """ Singleton object
    Mixer object to manage audio drivers 
    """
    __single_instance = None
    @staticmethod
    def get_instance():
        """ Static access method """
        if AudiMixer.__single_instance is None:
            AudiMixer()
        return AudiMixer.__single_instance

    def __init__(self, audio_driver=None):
        if AudiMixer.__single_instance != None:
            raise exception("This Klass is a singleton klass.")
        else:
            AudiMixer.__single_instance = self
        
        self.audio_driver = None
        # self._audio_driver = AlsaAudioDriver()
        if audio_driver is None or audio_driver == "port":
            self.audio_driver = aupor.AudiPortDriver()
        # must be the same as buf_size in PortAudio Driver
        self._buf_size =512
        self._channels =1 # TODO: why nchannels not been initialized by init function?
        self._rate = 44100
        self._format =0
    #-----------------------------------------

    def init(self, channels=2, rate=44100, format=None, input_device_index=None, output_device_index=None):
        if self.audio_driver is None: return
        
        self._channels = channels
        self._rate = rate
        if format is None: format = self._format
        else: self._format = format
        
        # self.audio_driver.set_callback(self._audio_callback)
        self.audio_driver.init_params(channels, rate, format)
        self.audio_driver.init_devices(input_device_index, output_device_index)

    #-----------------------------------------

    def set_audio_callback(self, audio_callback):
        if self.audio_driver is None: return
        self.audio_driver.set_audio_callback(audio_callback)

    #-----------------------------------------

    def open(self):
        if self.audio_driver is None: return
        self.audio_driver.open()

    #-----------------------------------------

    def start_driver(self):
        if self.audio_driver is None: return
        self.audio_driver.start_engine()
        beep()

    #-------------------------------------------
    
    def stop_driver(self):
        if self.audio_driver is None: return
        self.audio_driver.stop_engine()
        beep()

    #-------------------------------------------
     
    def close(self):
        """ Close the mixer """
        if self.audio_driver is None: return
        self.audio_driver.stop_engine()
        self.audio_driver.close()

    #-------------------------------------------

    def is_running(self):
        if self.audio_driver is None: return
        return self.audio_driver.is_running()

    #-------------------------------------------

    def print_devices(self):
        """
        display devices info
        from Mixer object
        """

        if self.audio_driver is None: return
        self.audio_driver.print_devices()

    #-------------------------------------------


#========================================


if __name__ == "__main__":
    app = AudiMixer()
    app.init()
    input("It's OK...")

#-----------------------------------------


src/auditools.py
#! /usr/bin/python3

"""
    File: auditools.py
    Audio tools functions for AudiNotes.
    Date: Mon, 13/02/2023
    Author: Coolbrother
"""
import numpy as np

def _mid2freq(note):
    """
    return freq number from note midi number
    """

    return 440.0 * pow(2, (note - 69) / 12.0)

#-------------------------------------------


def _gen_sine_table(freq=440, rate=44100, channels=1, _len=5):
    """ returns array of sine wave """
    # Note IMPORTANT: dont forget to enclose in parenthesis (rate * channels)
    # because I searched the problem during a whole night
    incr = (2 * np.pi * freq ) / (rate * channels)
    nbsamples = _len * rate * channels
    x = np.arange(nbsamples) # in dtype float64
    arr = np.sin(incr * x)
    
    """
    Note: to convert array mono to interleave stereo array
    _arr = np.zeros((arr.size * _channels), arr.dtype)
    _arr[0::2] = arr
    _arr[1::2] = arr
    """

    """
    Note: to convert 1D array stereo to 2D array stereo 
    like (in_data) array in recording callback
    result = np.frombuffer(in_data, dtype=np.float32)
    result = np.reshape(result, (frames_per_buffer, 2))
    # to access to the channels use:
    left_channel = result[:, 0]
    right_channel = result[:, 1]
    """

    # Note: IMPORTANT to convert it in the right format for pyaudio module
    # _arr = np.float32(_arr) 
    
    return np.float32(arr)

#-------------------------------------------

def _gen_notes(note_lst, _len=0.5, count=1, rate=44100, channels=2):
    """
    generate notes for tests
    """

    arr_lst = []
    
    """
    # C game
    note_lst = [
            60, 62, 64, 65, 
            67, 69, 71, 72,
            72, 74, 76, 77, 
            79, 81, 83, 84
            ]
    """

    for _ in range(count):
        for note in note_lst:
            freq = _mid2freq(note)
            arr = _gen_sine_table(freq=freq, _len=_len, rate=rate, channels=channels)
            arr_lst.append(arr)
        
    arr = np.concatenate(arr_lst)
    return arr

#-------------------------------------------

def _gen_array(arr, num=1, channels=1):
    """ 
    Note: deprecated function
    -- generator for returning slice of numpy array 
    """
    
    for i in range(arr.shape[0] // num * channels):
        yield arr[num*i:num*(i+1)]

#-------------------------------------------


def replace_data(play_data, rec_data):
    """
    returns replacing array by another
    from AudiTools object
    """
    
    play_len = play_data.size
    rec_len = rec_data.size
    if play_len < rec_len:
        # Just replacing play_data by rec_data
        return rec_data
    elif play_len >= rec_len:
        # take play_len as longest track
        final_data = np.zeros(play_len, dtype=np.float32)
        # Adding rec_data to the final_data
        for i in range(rec_len):
            final_data[i] = rec_data[i]
        # Adding the rest for play_data to the final_data
        for i in range(rec_len, play_len):
            final_data[i] = play_data[i]
        
        return final_data

#-------------------------------------------

def merge_data(play_data, rec_data, vol):
    """
    returns array with merging arrays together
    from AudiPlayer object
    """
    
    play_len = play_data.size
    rec_len = rec_data.size
    final_data = np.array([])
    if play_len < rec_len:
        # take rec_len as longest track
        final_data = np.zeros(rec_len, dtype=np.float32)
        # Adding play_data + rec_data to the final_data
        for i in range(play_len):
            final_data[i] = (play_data[i] + rec_data[i]) * vol
        # Adding the rest of rec_data to the final_data
        for i in range(play_len, rec_len):
            final_data[i] = rec_data[i] * vol
    elif play_len >= rec_len:
        # take play_len as longest track
        final_data = np.zeros(play_len, dtype=np.float32)
        # Adding rec_data + play_data to the final_data
        for i in range(rec_len):
            final_data[i] = (rec_data[i] + play_data[i]) * vol
        # Adding the rest of play_data to the final_data
        for i in range(rec_len, play_len):
            final_data[i] = play_data[i] * vol
    
    return final_data

#-------------------------------------------




src/auditrack.py
#! /usr/bin/python3

"""
    File: auditrack.py
    Lite Track klass for AudiNotes.
    Date: Mon, 13/02/2023
    Author: Coolbrother
"""
import numpy as np

class AudiTrack(object):
    """ Track audio object """
    def __init__(self):
        self.id = id
        self._active =0
        self._muted =0
        self._arm_muted =0
        self._buf_arr = np.array([], dtype=np.float32)
        self._pos =0
        self._len =0
        self._looping =0


    #-------------------------------------------

    def get_data(self):
        """
        returns numpy array
        """

        return self._buf_arr

    #-------------------------------------------

    def set_data(self, arr):
        """
        set data to the track
        from Track object
        """

        if arr is not None:
            self._buf_arr = arr
            self._len = self._buf_arr.size
            # self._pos =0

    #-----------------------------------------
    
    def get_length(self, unit=0):
        """
        return total track length 
        from track object
        """

        return self._len
    
    #-----------------------------------------
    
    def init_position(self):
        """
        init track position 
        from track object
        """
        
        self.set_position(0, 0) # in frames
        
    #-----------------------------------------

    def get_position(self, unit=0):
        """
        return track position
        from Track object
        """

        return self._pos

    #-----------------------------------------

    def set_position(self, pos, unit=0):
        """
        set track position 
        from track object
        """

        if pos <0: pos =0
        elif pos > self._len: pos = self._len
        
        self._pos = pos

    #-----------------------------------------

    def is_active(self):
        """
        returns active state for this track
        from Track object
        """
        
        return self._active

    #-----------------------------------------

    def set_active(self, active):
        """ 
        set active state for this track
        from Track object
        """

        self._active = active

    #-----------------------------------------

    def get_mute(self):
        return (self._leftmute, self._rightmute)
    
    #-----------------------------------------

    def is_muted(self):
        return self._muted
    
    #-----------------------------------------
 
    def set_muted(self, muted=0, chanside=2):
        # set channel mute
        val =0 if muted else 1 # ternary operator
        if chanside == 0: # left channel side
            self._leftmute = val
        elif chanside == 1: # right channel side
            self._rightmute = val
        elif chanside == 2: # both channel side
            self._leftmute = val
            self._rightmute = val

        self._muted = muted

    #-----------------------------------------

    def toggle_mute(self):
        """
        toggle muted state
        from AudiTrack object
        """

        self._muted = not self._muted
        
        return self._muted

    #-----------------------------------------

    def is_arm_muted(self):
        """
        return internal mute state for armed track
        from Track object
        """
        
        return self._arm_muted

    #-----------------------------------------

    def set_arm_muted(self, arm_muted=0):
        """
        set internal mute state for armed track
        from Track object
        """
        
        self._arm_muted  = arm_muted

    #-----------------------------------------

    def set_looping(self, looping):
        """
        set looping state
        from AudiTrack object
        """

        self._looping = looping

    #-----------------------------------------

    def is_looping(self):
        """
        returns looping state
        from AudiTrack object
        """

        return self._looping

    #-----------------------------------------

    def toggle_loop(self):
        """
        toggle looping state
        from AudiTrack object
        """

        self._looping = not self._looping
        
        return self._looping

    #-----------------------------------------


    def write_sound_data(self, out_data, data_count):
        """
        writing curdata to the out_data
        from AudiTrack object
        """

        vol =0.5
        curdata = self._buf_arr
        if not curdata.size: return
        pos = self._pos
        _len = self._len
        
        if self._muted or self._arm_muted:
            # dont write any data, just increment curpos
            for i in range(0, data_count, 2):
                if pos +1 >= _len: # End of buffer
                    if self._looping:
                        pos =0
                else: # pos < _len
                    pos += 2

        else: # not muted, not armed
            for i in range(0, data_count, 2):
                if pos +1 >= _len: # End of buffer
                    if self._looping:
                        pos =0
                else: # pos < _len
                    # attenuate amplitude data before adding it, cause others data are allready attenuated
                    val = curdata[pos] * vol
                    out_data[i] = (out_data[i] + val)
                    out_data[i+1] = (out_data[i+1] + val)
                    pos += 2
        self._pos = pos
    
    #-----------------------------------------

#========================================

if __name__ == "__main__":
    tra = AudiTrack()
    _len = tra._len
    print(f"Track len: {_len}")
    input("It's OK...")
#-----------------------------------------


src/portdriver.py
#!/usr/bin/python3 
"""
    File: portdriver.py
    portdriver module
    Portaudio module with pyaudio

    Last update: Wed, 01/02/2023
    Version: 0.2
    -- Adding functions:
    is_running, query_devices, print_devices


    
    Date: Tue, 31/01/2023
    Author: Coolbrother
"""
import pyaudio

class PortDriver(object):
    """ PortAudio driver manager  """
    def __init__(self):
        self._pa = pyaudio.PyAudio()
        self._stream = None
        self._buf_size = 1024
        self._rate =44100
        self._channels =2 # FIX: work fine with channels 2
        # Format work with paInt16 or paFloat32, for default Soundcard.
        # self._format = pyaudio.paInt16
        self._format = pyaudio.paFloat32
        self._input_device_index = None
        self._output_device_index = None
        self._default_input_index = None
        self._default_output_index = None
        self._callback_func = None
        self._running = False

    #-------------------------------------------

    def open(self):
        try:
            self._stream = self._pa.open(
                            format = self._format, 
                            channels = self._channels, 
                            rate = self._rate, # wf.getframerate(),
                            input_device_index = self._input_device_index,
                            output_device_index = self._output_device_index,
                            input = True,
                            output = True,
                            stream_callback = self._callback_func,
                            start = False,
                            frames_per_buffer = self._buf_size
                            )
        except OSError as err:
            print("[PortAudio Error]: error to open input or output device index", err)
            self._stream = None

        if self._stream is None:
            self._output_device_index = self._default_output_index
            self._input_device_index = self._default_input_index
            try:
                self._stream = self._pa.open(
                                format = self._format, 
                                channels = self._channels, 
                                rate = self._rate, # wf.getframerate(),
                                input_device_index = self._input_device_index,
                                output_device_index = self._output_device_index,
                                input = True,
                                output = True,
                                stream_callback = self._callback_func,
                                start = False,
                                frames_per_buffer = self._buf_size
                                )
            except OSError as err:
                print("[PortAudio Error]: error to open input or output device index", err)
                self._stream = None

    #-----------------------------------------

    def start_engine(self):
        if self._stream and not self._running:
            self._stream.start_stream()
            print("Starting Engine...")
            self._running = True
        
    #-----------------------------------------

    def stop_engine(self):
        self._running = False
        if self._stream:
            self._stream.stop_stream()
            print("Stopping Engine...")

    #-----------------------------------------

    def close(self):    
        if self._stream:
            self._stream.close()
            print("Closing the Stream")
        
        self._pa.terminate()
        print("Closing the driver")

    #-----------------------------------------

    def init_devices(self, input_device_index=None, output_device_index=None):
        """ initialize default devices index """
        self._input_device_index = input_device_index
        self._output_device_index = output_device_index
        self._default_input_index = self._pa.get_default_input_device_info().get('defaultInputDevice')
        self._default_output_index = self._pa.get_default_output_device_info().get('defaultOutputDevice')

    #-------------------------------------------

    def is_running(self):
        return self._running

    #-------------------------------------------

    def set_audio_callback(self, audio_callback):
        self._callback_func = audio_callback

    #-------------------------------------------

    def get_driver_names(self):
        """ return list of host api names
        """

        lst = []
        count =0
        count = self._pa.get_host_api_count()
        for i in range(count):
            dic = self._pa.get_host_api_info_by_index(i)
            try:
                    name = dic.get('name')
                    lst.append((i, name))
            except KeyError:
                pass

        return lst

    #-----------------------------------------
       
    def get_device_list(self):
        """ return list containing all devices info in a dictionary
        """
        lst = []
        count = self._pa.get_device_count()
        for i in range(count):
            lst.append(self._pa.get_device_info_by_index(i))
        
        return lst
    #-----------------------------------------

    def query_devices(self):
        """
        Returns list of dict with devices info
        from PortAudioDriver object
        """
        
        lst = []
        for  (index, item) in enumerate(self.get_device_list()):
            mark = ""
            host_index = item['hostApi']
            host_name = self.get_driver_names()[host_index][1]
            if item['name'] == "default": mark = "*"
            dic = {}
            dic['index'] = f"{mark} {item['index']}" 
            dic['name'] = f"{item['name']}"
            dic['hostName'] = f"{host_name}" 
            dic['maxInputChannels'] = f"{item['maxInputChannels']} In"
            dic['maxOutputChannels'] = f"{item['maxOutputChannels']} Out"
            lst.append(dic.copy())

            """
            print(f"{mark} {item['index']} {item['name']},", 
                f"{host_name} ({item['maxInputChannels']} In, {item['maxOutputChannels']} Out)")
            """
            
        return lst
    #-----------------------------------------

    def print_devices(self):
        print("Devices list")
        dic_lst = self.query_devices()
        for dic in dic_lst:
            index = dic['index']
            name = dic['name']
            hostName = dic['hostName']
            maxInputChannels = dic['maxInputChannels']
            maxOutputChannels = dic['maxOutputChannels']
            print(f"{index} {name},", \
                    f"{hostName} ({maxInputChannels}, {maxOutputChannels}")

    #-----------------------------------------


    def get_default_input_device(self):
        """ return tuple containing indexes input for both channels  
        """

        return self._pa.get_default_input_device_info().get('defaultInputDevice')

    #-----------------------------------------

    def get_default_output_device(self):
        """ return tuple containing indexes output for both channels  
        """

        return _pa.get_default_input_device_info().get('defaultOutputDevice')

    #-----------------------------------------

#========================================

def main():
    _audio_driver = PortDriver()
    _audio_driver.print_devices()

#-----------------------------------------

if __name__ == "__main__":
    main()
    input("Tapez Enter...")
#-----------------------------------------


src/audinotes.py
#! /usr/bin/python
"""
    File: audinotes.py
	Test unlimited recording on PyAudio
    Inspired from playrec module.
    Note: Work fine.
    with: output_device_index =15: default output Soundcard with only output
    input_device_index =0: default Input Soundcard, with only input
    output_device_index =6: Eternal output Soundcard
    input_device_index =6: External input Soundcard
    
    Last update: Fri, 03/02/2023
    See changelog

   
    Last update: Wed, 01/02/2023
    Version: 0.1
    
	Date: samedi, 26/07/14 02/43/56
    Author: Coolbrother

"""
#-------------------------------------------

import os, sys, time
import readline
import audiplayer as aupla
import audimixer as aumix

_DEBUG =1
_help = """ Help on AudiNotes Player
  b: forward
  k: toggle click
  h, ?: print this help
  l: toggle loop
  p, t, space: toggle play pause
  q, Q: quit
  r: toggle record
  R: toggle record mode (replace, mix)
  v: stop
  w: rewind
  x: toggle mute
  z: wiring
  <: goto start
  >: goto end

  bpm VAL: set bpm
  dev: display devices infomations
  eng, engine on/off: toggle audio engine
  init: init track
  sta, status: display player status and position in secs
  test: testing

"""

def debug(msg="", title="", bell=True):
    if _DEBUG:
        if title: msg = f"{title}: {msg}"
        print(msg)
        if bell: print("\a")
    
#------------------------------------------------------------------------------

def beep():
    print("\a")

#-------------------------------------------

class MainApp(object):
    """ Main application manager """
    def __init__(self):
        self.mixer = None
        self.player = None
        self.audio_driver = None

    #------------------------------------------------------------------------------

    def clear_screen(self, numlines=100):
        """
        simply clear the screen for Windows or Linux plateform
        from MainApp object
        """

        os.system('cls||clear')

    #------------------------------------------------------------------------------

    def display(self, msg):
        """
        display message
        from MainApp object
        """

        print(msg)

    #------------------------------------------------------------------------------

    def init_app(self, input_device_index, output_device_index):
        self.mixer = aumix.AudiMixer()
        self.mixer.init(channels=1, rate=44100, format=None, input_device_index=input_device_index, output_device_index=output_device_index)
        self.audio_driver = self.mixer.audio_driver

        self.player = aupla.AudiPlayer(parent=self)
        self.player.init_player(self.mixer)
        # pl.start_engine()
        self.clear_screen()

    #------------------------------------------------------------------------------

    def change_engine(self, param=None):
        if param is None: return
        if param == "on":
            self.mixer.start_driver()
        elif param == "off":
            self.mixer.stop_driver()

    #-------------------------------------------

    def print_devices(self):
        """
        display devices info
        from MainApp object
        """

        self.mixer.print_devices()

    #-------------------------------------------

    def change_bpm(self, bpm, adding=0):
        if self.mixer is None: return
        cur_bpm = self.player.get_bpm()
        if type(bpm) == str:
            if bpm[0] in ('+', '-'): adding=1
        bpm = float(bpm)
        if adding == 1: # is incremental
            bpm += cur_bpm

        cur_bpm = self.player.set_bpm(bpm)
        msg = f"Bpm: {cur_bpm}"
        self.display(msg)

    #-------------------------------------------


    def main(self, input_device_index, output_device_index):
        self.init_app(input_device_index, output_device_index)
        msg = "Press '?' or 'h' for help"
        
        self.display(msg)
        sav_str = ''

        try:
            while 1:

                key = param1 = param2 = param3 = None
                val_str = input("-> ")
                if val_str == '': val_str = sav_str
                else: sav_str = val_str

                if val_str == " ":
                    key = val_str
                else:
                    lst = val_str.split()
                    len_lst = len(lst)
                    if len_lst >0: key = lst[0]
                    if len_lst >1: param1 = lst[1]
                    if len_lst >2: param2 = lst[2]
                    if len_lst >3: param3 = lst[3]

                
                if key in ('q', 'Q'):
                    self.player.close()
                    self.display("Quit...")
                    break

                elif key == 'b': # Forward
                    pos = self.player.forward()
                    pos = self.player.samples_to_sec(pos)
                    msg = f"Time at: {pos:.3f} Secs"
                    self.display(msg)
                elif key in (' ', 't', 'p'):
                    if not self.player.is_playing():
                        msg = "Play"
                        self.player.play()
                    else:
                        self.player.pause()
                        pos = self.player.get_position()
                        pos = self.player.samples_to_sec(pos) # convert to sec
                        msg = f"Pause at: {pos:.3f} secs"
                    self.display(msg)
                elif key == 'k':
                    val = self.player.toggle_click()
                    if val: msg = "Start Clicking"
                    else: msg = "Stop Clicking"
                    self.display(msg)
                elif key == 'l':
                    val = self.player.toggle_loop()
                    if val: msg = "Start Looping"
                    else: msg = "Stop Looping"
                    self.display(msg)
                elif key == 'r':
                    # toggle record
                    val = self.player.toggle_record()
                    if val: msg = "Start Record"
                    else: msg = "Stop Record"
                    self.display(msg)

                elif key == 'R':
                    # toggle record mode
                    val = self.player.toggle_rec_mode()
                    if val == 0:
                        msg = "Record mode Replace"
                    else:
                        msg = "Record mode Mix"
                    self.display(msg)
                elif key == 'v':
                    self.player.stop()
                    pos = self.player.get_position()
                    pos = self.player.samples_to_sec(pos)
                    msg = f"Stop at: {pos:.3f} Secs"
                    self.display(msg)
                elif key == 'w': # Rewind
                    pos = self.player.rewind()
                    pos = self.player.samples_to_sec(pos)
                    msg = f"Time at: {pos:.3f} Secs"
                    self.display(msg)
                elif key == 'x':
                    val = self.player.toggle_mute()
                    if val: msg = "Muted On"
                    else: msg = "Muted Off"
                    self.display(msg)
                elif key == 'z': # wiring
                    self.player.recwire()
                elif key == '<': # goto start
                    pos = self.player.goto_start()
                    pos = self.player.samples_to_sec(pos)
                    msg = f"Goto Start at: {pos:.3f} Secs"
                    self.display(msg)
                elif key == '>': # goto End
                    pos = self.player.goto_end()
                    pos = self.player.samples_to_sec(pos)
                    msg = f"Goto End at: {pos:.3f} Secs"
                    self.display(msg)
                elif key in ('?', 'h',):
                    self.display(_help)

                elif key == 'bpm':
                    # change bpm
                    if not param1: param1 = "120"
                    self.change_bpm(param1, adding=0) # can be incremental or decremental with (+, -)
                    
                elif key == 'dev':
                    self.print_devices()
                elif key in ('eng', 'engine'):
                    self.change_engine(param1)
                elif key == 'init':
                    self.player.init_track()
                elif  key in ('sta', 'status'): # Status
                    state = self.player.get_state()
                    pos = self.player.get_position()
                    pos = self.player.samples_to_sec(pos)
                    # start_loop = self.player.get_start_loop()
                    # end_loop = self.player.get_end_loop()
                    msg = f"{state}, Position: {pos:.3f} Secs"
                    self.display(msg)
                elif key == 'test':
                    self.test()
                 
                else:
                    msg = "Command not found"
                    self.display(msg)
        finally:
            # save history file
            pass

    #----------------------------------------

    def test(self):
        """
        testing function
        from MainApp object
        """
        
        self.display("Testing...")

        lat_ms = self.player._input_latency + self.player._output_latency # latency in msec
        rate = self.player._rate
        channels = self.player._channels
        shift_samples = int(lat_ms * rate * channels) # in samples
        print(f"lat_ms: {lat_ms}, rate: {rate}, channels: {channels}")
        print(f"Shift Samples: {shift_samples}")
        # self.player.init_track()

    #----------------------------------------

#========================================


if __name__ == "__main__":
    input_device_index =6 # External Soundcard
    # output_device_index =5 # Default output Soundcard
    output_device_index =6 # External Soundcard
    app = MainApp()
    app.main(input_device_index, output_device_index)
#----------------------------------------
